### PR TITLE
Refactor proxy database resource records

### PR DESCRIPTION
**Pull request by gabrieljackson**
This refactor takes logical database and database schema information
that was previously stored in the root multitenant database object
and splits it out into new tables. This is done to create a more
flexible design that will allow the provisioner to store more
metadata about each resource type.

Fixes https://mattermost.atlassian.net/browse/MM-37442

Reviewer notes:

Although there are now new resource types for logical
databases and database schemas, these are similar to cluster
installations in that they are only created and deleted as part of
installation creation and deletion. APIs for listing these resources
will come in a second PR as this one was already very large.

```release-note
Refactor proxy database resource records
```

### CODE REVIEW

**Review by Szymongib — APPROVED** [review]
Looks good!

No major concerns from my side.

File: internal/store/db_logical_database.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+// CreateLogicalDatabase records the supplied logical database to the datastore.
+func (sqlStore *SQLStore) CreateLogicalDatabase(logicalDatabase *model.LogicalDatabase) error {
+	multitenantDatabase, err := sqlStore.GetMultitenantDatabase(logicalDatabase.MultitenantDatabaseID)
+	if err != nil {
+		return errors.Wrap(err, "failed to get multitenant database")
+	}
+	if multitenantDatabase == nil {
+		return errors.Errorf("multitenant database %s does not exist", logicalDatabase.MultitenantDatabaseID)
+	}
```
[the remark this answers is not in the document]
Comment (gabrieljackson) [reply internal/store/db_logical_database.go]: Yep, I put this in originally to protect things while I prototyped, but just removed it to standardize. Thanks!

File: internal/store/migrations.go [hunk 2]
```diff
@@ -1480,6 +1483,169 @@ var migrations = []migration{
...
+			FROM
+			MultitenantDatabaseBackup2;
+		`)
+		if err != nil {
+			return err
+		}
+
+		// The MultitenantDatabaseBackup2 table will be left for now just in case
+		// and will be cleaned up in a future migration along with the previous
+		// MultitenantDatabaseBackup table.
```
[the remark this answers is not in the document]
Comment (gabrieljackson) [reply internal/store/migrations.go:1647]: Sure thing!

File: model/multitenant_database.go [hunk 3]
```diff
@@ -22,45 +21,42 @@ type MultitenantDatabase struct {
...
-	}
+// LogicalDatabase represents a logical database inside a MultitenantDatabase.
+type LogicalDatabase struct {
+	ID                    string
+	MultitenantDatabaseID string
+	Name                  string
+	CreateAt              int64
+	DeleteAt              int64
+	LockAcquiredBy        *string
+	LockAcquiredAt        int64
```
[the remark this answers is not in the document]
Comment (gabrieljackson) [reply model/multitenant_database.go:39]: I do have plans for how we could use these including migration logic so I added them here although they are not yet used. I consider the migration logic to be somewhat expensive even if it is just for unit tests so I would encourage us to plan a bit ahead if we can to reduce the number of times we do it.

**Comment by gabrieljackson** [comment]
It's important to get this soaking in our test account so going to merge what we have and will address issues in a follow-up PR.

**Pull request merged**